internal/controllers/synthesis/lifecycle.go

```diff
@@ -392,10 +393,46 @@ func shouldSwapStates(synth *apiv1.Synthesizer, comp *apiv1.Composition) bool {
 	// - synthesis is not already pending
 	// - all bound input resources exist and are in lockstep (or composition is being deleted)
 	syn := comp.Status.CurrentSynthesis
-	return (syn == nil ||
-		syn.ObservedCompositionGeneration != comp.Generation ||
-		(!inputRevisionsEqual(synth, comp.Status.InputRevisions, syn.InputRevisions) && syn.Synthesized != nil && !comp.ShouldIgnoreSideEffects())) &&
-		(comp.DeletionTimestamp != nil || (comp.InputsExist(synth) && !comp.InputsOutOfLockstep(synth)))
+
+	isSynNil := syn == nil
+	isCompGenDiff := syn != nil && syn.ObservedCompositionGeneration != comp.Generation
+	isInputRevisionsEqual := syn != nil && inputRevisionsEqual(synth, comp.Status.InputRevisions, syn.InputRevisions)
+	isSynthesizedNotNil := syn != nil && syn.Synthesized != nil
+	shouldIgnoreSideEffects := comp.ShouldIgnoreSideEffects()
+	isCompDeleted := comp.DeletionTimestamp != nil
+	isCompInputsExist := comp.InputsExist(synth)
+	isCompInputsOutOfLockstep := comp.InputsOutOfLockstep(synth)
+
+	reason := ""
+	if isSynNil {
+		reason += "current synthesis is nil"
+	} else {
+		reason += "current synthesis is not nil"
+		if isCompGenDiff {
+			reason += ", observed composition generation is different"
+		}
+		if !isInputRevisionsEqual && isSynthesizedNotNil && !shouldIgnoreSideEffects {
+			reason += ", input revisions are not equal and synthesized is not nil and should not ignore side effects"
+		}
+	}
+
+	if isCompDeleted {
+		reason += ", composition is deleted"
+	}
+	if isCompInputsExist && !isCompInputsOutOfLockstep {
+		reason += ", composition inputs exist and inputs are not out of lockstep"
+	}
+
+	res := (isSynNil ||
+		isCompGenDiff ||
+		(!isInputRevisionsEqual && isSynthesizedNotNil && !shouldIgnoreSideEffects)) &&
+		(isCompDeleted || (isCompInputsExist && !isCompInputsOutOfLockstep))
+
+	if res {
+		logger = logger.WithValues("swap composition synthesis reason", reason)
+	}
+
+	return logger, res
```

Comment:
nit: Consider returning a string instead of wrapping the logger, easier to unit test

Reply:
done